apps/openmw/mwworld/worldimp.cpp

```diff
@@ -2681,18 +2684,30 @@ namespace MWWorld
         if (result2.mHit)
             dist2 = (origin - result2.mHitPointWorld).length();
 
-        if (dist1 <= dist2 && result1.mHit)
+        if (result1.mHit)
         {
             target = result1.mHitObject;
             hitPosition = result1.mHitPos;
             if (dist1 > getMaxActivationDistance() && !target.isEmpty() && (target.getClass().isActor() || !target.getClass().canBeActivated(target)))
                 target = NULL;
+            if (!Settings::Manager::getBool("allow AI to hit unintended targets", "Game"))
+            {
+                // Don't let AI actors hit unintended targets. If an unintended target is hit, continue casting the ray past them
+                while (actor != MWMechanics::getPlayer() && !target.isEmpty() && target.getClass().isActor() && !stats.getAiSequence().isInCombat(target))
+                {
+                    result1 = mPhysics->castRay(result1.mHitPos, dest, target, MWPhysics::CollisionType_Actor);
+                    if (result1.mHit)
+                        target = result1.mHitObject;
+                    else
+                        target = NULL;
```

Comment:
I thought I had accounted for the infinite loop (I actually wrote this a while ago, not today), but maybe not.

> I would use a single cast and filter unintended targets via the ResultCallback.

I'm not quite sure what that involves at the moment but I'll see if I can figure it out.

> I'd remove the setting for now; really no reason to enable at the moment with the AI being as bad as it is.

I thought people might want it anyway but I'll go ahead and remove it then. I agree it's really a problem at the moment and I'm kind of surprised I haven't seen it brought up more.